packages/react-components/src/components/ComplianceBanner/ComplianceBanner.tsx

```diff
@@ -147,6 +152,88 @@ function determineStates(previous: ComplianceState, current: boolean | undefined
   }
 }
 
+interface TimestampedVariant {
+  variant: ComplianceBannerVariant;
+  // Milliseconds since epoch (i.e., return value of Date.now()).
+  lastUpdated: number;
+}
+
+/**
+ * Shows a {@link BannerMessage} in a {@link MessageBar} tracking `variant` internally.
+ *
+ * This component delays and combines frequent updates to `variant` such that:
+ * - Updates that happen within {@link BANNER_OVERWRITE_DELAY_MS} are delayed.
+ * - Once {@link BANNER_OVERWRITE_DELAY_MS} has passed since the last update, the _latest_ pending update is shown.
+ *
+ * This ensures that there is enough time for the user to see a banner message before it is overwritten.
+ * In case of multiple delayed messages, the user always sees the final message as it reflects the final state
+ * of recording and transcription.
+ *
+ * @private
+ */
+function DelayedUpdateBanner(props: {
+  variant: TimestampedVariant;
+  onDismiss: () => void;
+  strings: _ComplianceBannerStrings;
+}): JSX.Element {
+  const { variant, lastUpdated: variantLastUpdated } = props.variant;
+
+  // Tracks the variant that is currently visible in the UI.
+  const [visible, setVisible] = useState<TimestampedVariant>({
+    variant,
+    lastUpdated: Date.now()
+  });
+  const pendingUpdateHandle = useRef<ReturnType<typeof setTimeout> | null>(null);
+
+  if (variant !== visible.variant && variantLastUpdated > visible.lastUpdated) {
+    // Always clear pending updates.
+    // We'll either update now, or schedule an update for later.
+    if (pendingUpdateHandle.current) {
+      clearTimeout(pendingUpdateHandle.current);
```

Comment:
Must set `pendingUpdateHandle.current` to `null` again.

`clearTimeout` is forgiving so this doesn't cause errors in practice, but still.